Spearman cor test

mapi1 opened the pull request:
This PR adds the `SpearmanCorrelationTest` as suggested in #236. 
For the confidence interval I took inspiration from [this](https://stats.stackexchange.com/questions/18887/how-to-calculate-a-confidence-interval-for-spearmans-rank-correlation)  StackExchange thread and used the suggested variance estimator to counter the non-normal distribution of the ranks.
Unfortunately, I could not really add meaningful tests for it as R's `cor.test` does not give the intervals for Spearman correlation and uses another algorithm to calculate the p-value as well. Maybe someone has an idea here or knows a tool that can calculate this already correctly.

nalimilan commented on src/correlation.jl:
```diff
@@ -2,8 +2,11 @@
+abstract type AbstractCorrelationTest <: HypothesisTest end
 
 """
     CorrelationTest(x, y)
```
Add a mention of Pearson to this docstring to differentiate it from the new test.

nalimilan commented on src/correlation.jl:
```diff
@@ -80,12 +84,65 @@ function StatsAPI.confint(test::CorrelationTest{T}, level::Float64=0.95) where T
+"""
+    SpearmanCorrelationTest(x, y)
+
+Perform a t-test for the hypothesis that ``\\text{Cor}(x,y) = 0``, i.e. the rank-based Spearman correlation 
```
Break all lines at 92 chars like in the `CorrelationTest` docstring. Also I would say "Spearman rank correlation" rather than "rank-based".

nalimilan commented on src/correlation.jl:
```diff
@@ -80,12 +84,65 @@ function StatsAPI.confint(test::CorrelationTest{T}, level::Float64=0.95) where T
+    end
+end
+
+testname(p::SpearmanCorrelationTest) =  "Spearman correlation"
```
```suggestion
testname(p::SpearmanCorrelationTest) = "Spearman correlation"
```

nalimilan commented on src/correlation.jl:
```diff
@@ -2,8 +2,11 @@
+using StatsBase: corspearman
 
-export CorrelationTest
+export CorrelationTest, SpearmanCorrelationTest
```
Also add new test to docs/src.

nalimilan commented on test/correlation.jl:
```diff
@@ -26,6 +26,30 @@ using DelimitedFiles
+    let out = sprint(show, w)
+        @test occursin("reject h_0", out) && !occursin("fail to", out)
+    end
+    # let ci = confint(w)
```
Why is this commented out?

nalimilan commented on test/correlation.jl:
```diff
@@ -26,6 +26,30 @@ using DelimitedFiles
+    #     @test first(ci) ≈ -0.1105478 atol=1e-6
+    #     @test last(ci) ≈ 0.0336730 atol=1e-6
+    # end
+    @test pvalue(x) ≈ 0.09275 atol=1e-2 # value from R's cor.test(..., method="spearman") which does not use a t test algorithm AS 89 
```
Can we find a more precise value to test against? This way of writing the test is misleading as even `0.09` would pass.

In the worst case, we should test with a lower tolerance against the value we return, and just note in a comment the value returned by R.

nalimilan commented on src/correlation.jl:
```diff
@@ -80,12 +84,65 @@ function StatsAPI.confint(test::CorrelationTest{T}, level::Float64=0.95) where T
+
+Implements `pvalue` for the t-test.
+
+Implements `confint` using an approximate confidence interval adjusting for the non-normality of the ranks based on [1]. This is still an approximation and which performs insufficient in the case of:
+
+* small sample sizes n < 25
```
```suggestion
Implements `confint` using an approximate confidence interval adjusting for the non-normality of the ranks based on [1]. This is still an approximation, which performs insufficiently in the case of:
* sample sizes below 25
```

nalimilan commented on src/correlation.jl:
```diff
@@ -80,12 +84,65 @@ function StatsAPI.confint(test::CorrelationTest{T}, level::Float64=0.95) where T
+[1] D. G. Bonett and T. A. Wright, “Sample size requirements for estimating pearson, kendall and spearman correlations,” Psychometrika, vol. 65, no. 1, pp. 23–28, Mar. 2000, doi: 10.1007/BF02294183.
+
+[2] A. J. Bishara and J. B. Hittner, “Confidence intervals for correlations when data are not normal,” Behav Res, vol. 49, no. 1, pp. 294–309, Feb. 2017, doi: 10.3758/s13428-016-0702-8.
+
```
```suggestion
```

nalimilan commented on src/correlation.jl:
```diff
@@ -80,12 +84,65 @@ function StatsAPI.confint(test::CorrelationTest{T}, level::Float64=0.95) where T
+Implements `confint` using an approximate confidence interval adjusting for the non-normality of the ranks based on [1]. This is still an approximation and which performs insufficient in the case of:
+
+* small sample sizes n < 25
+* a high true population Spearman correlation
```
According to the StackExchange thread this is more precisely:
```suggestion
* a true population Spearman correlation above 0.95
```
It also mentions ordinal data. Did you omit it on purpose? I admit it's not super explicit.

nalimilan commented on src/correlation.jl:
```diff
@@ -80,12 +84,65 @@ function StatsAPI.confint(test::CorrelationTest{T}, level::Float64=0.95) where T
+In these cases a bootstrap confidence interval can perform better [2].
+
+# External resources
+[1] D. G. Bonett and T. A. Wright, “Sample size requirements for estimating pearson, kendall and spearman correlations,” Psychometrika, vol. 65, no. 1, pp. 23–28, Mar. 2000, doi: 10.1007/BF02294183.
```
```suggestion
[1] D. G. Bonett and T. A. Wright, “Sample size requirements for estimating Pearson, Kendall and Spearman correlations,” Psychometrika, vol. 65, no. 1, pp. 23–28, Mar. 2000, doi: 10.1007/BF02294183.
```

mapi1 replied:
Thanks for your detailed review! I tried to incorporate it as suggested. 

I used the `spearmanCI` R library to get values for the CIs for testing. They suffer from the same problem as the p value, as they have a low number of matching significant digits.  I still wrote the tests as a form of documentation and also added tests to compare against the vales we return to catch feature changes/bugs etc.

Also mention now the ordinal data as it is the main message of the Ruscio paper (Now added to docstring).

nalimilan reviewed (commented):
Thanks, looks mostly good!

I'm not sure which existing implementations could be used to check results. Have you looked at those mentioned by Wikipedia, as there are several which return p-values?

mapi1 commented:
Thanks for your detailed review! I tried to incorporate it as suggested. 

I used the `spearmanCI` R library to get values for the CIs for testing. They suffer from the same problem as the p value, as they have a low number of matching significant digits.  I still wrote the tests as a form of documentation and also added tests to compare against the vales we return to catch feature changes/bugs etc.

Also mention now the ordinal data as it is the main message of the Ruscio paper (Now added to docstring).

nalimilan reviewed (commented):
Sorry for the delay. Looks almost ready. I've just made a few more comments.

Regarding comparison of CIs against R, I hadn't realized spearmanCI doesn't implement the same CI method. In that case I don't think it makes sense to test against these values, as there are mathematically legitimate reasons to get different results. Maybe you could check against code that isn't included in a package such as [this one](https://rpubs.com/seriousstats/616206) instead? Then if that matches we can just test against the exact values we return to prevent regressions.

dpinol commented:
@mapi1 thanks for the PR. Do you foresee you'll have the opportunity to move it forward? cheers
